### PR TITLE
PRC-124 : Return contacts without address

### PR DESCRIPTION
**Pull request by prabash-moj**
Changes : 
1) Results include contacts without address
2) Add integration test
3) Refactor criteria query into a view for maintainability and handle Left Join  without using explicit relationships in entities

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppscontactsapi/entity/ContactAddressSummaryEntity.kt [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+import jakarta.persistence.Enumerated
+import jakarta.persistence.Id
+import jakarta.persistence.Table
+import uk.gov.justice.digital.hmpps.hmppscontactsapi.model.request.EstimatedIsOverEighteen
+import java.time.Instant
+import java.time.LocalDate
+import java.time.LocalDateTime
+
+@Entity
+@Table(name = "v_contact_address")
```
Comment (james-baker-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppscontactsapi/entity/ContactAddressSummaryEntity.kt:14]: "contact address" seems like a bad name for all of this. Especially seeing as the point is that they might not have an address.

File: src/main/resources/migrations/common/V2024.08.19.1__reset_sequences.sql [hunk 2]
```diff
@@ -2,12 +2,12 @@
...
 -- themselves need to be reset to the next number for generated ID values.
 --
-alter sequence if exists contact_contact_id_seq restart with 16;
-alter sequence if exists contact_address_contact_address_id_seq restart with 16;
-alter sequence if exists contact_email_contact_email_id_seq restart with 4;
-alter sequence if exists contact_identity_contact_identity_id_seq restart with 4;
-alter sequence if exists contact_phone_contact_phone_id_seq restart with 12;
-alter sequence if exists reference_codes_reference_code_id_seq restart with 82;
-alter sequence if exists prisoner_contact_prisoner_contact_id_seq restart with 29;
+alter sequence if exists contact_contact_id_seq restart with 17;
```
Comment (timharrison-moj) [inline src/main/resources/migrations/common/V2024.08.19.1__reset_sequences.sql:5]: You've added contact ID 17, so should start with 18 for the nextval?

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppscontactsapi/service/ContactServiceTest.kt [hunk 3]
```diff
@@ -388,21 +387,15 @@ class ContactServiceTest {
       assertThat(result.content[0].firstName).isEqualTo("first")
     }
 
-    private fun getContactEntity(contactId: Long) = ContactEntity(
-      contactId = contactId,
+    private fun getContactAddressSummaryEntity() = ContactSummaryEntity(
+      contactId = 1L,
```
Comment (timharrison-moj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppscontactsapi/service/ContactServiceTest.kt:391]: Not keen on the name `ContactSummaryEntity` - and the address join seems to spoil the purpose of the search ... to find contacts by partial names, DOb's etc,

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppscontactsapi/mapping/ContactsServiceMappers.kt [hunk 4]
```diff
@@ -27,35 +27,25 @@ fun ContactEntity.toModel() = Contact(
...
-    street = contactAddress.street,
-    area = contactAddress.area,
-    cityCode = contactAddress.cityCode,
-    countyCode = contactAddress.countyCode,
-    postCode = contactAddress.postCode,
-    countryCode = contactAddress.countryCode,
-  )
-}
+fun ContactSummaryEntity.toModel() = ContactSearch(
+  id = this.contactId,
```
Comment (timharrison-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppscontactsapi/mapping/ContactsServiceMappers.kt:31]: ContactWithAddressEntity,   ContactSearchResultItem?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppscontactsapi/entity/ContactSummaryEntity.kt [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+import jakarta.persistence.Table
+import uk.gov.justice.digital.hmpps.hmppscontactsapi.model.request.EstimatedIsOverEighteen
+import java.time.Instant
+import java.time.LocalDate
+import java.time.LocalDateTime
+
+@Entity
+@Table(name = "v_contact_summary")
+data class ContactSummaryEntity(
+
```
Comment (timharrison-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppscontactsapi/entity/ContactSummaryEntity.kt:16]: ContactWithAddressEntity?

**Review by timharrison-moj — COMMENTED** [review]
Small naming changes suggested for now - to get the first goal of finding contacts with or without an address.  Longer term, I think another ticket to separate contact search from enriching results with addresses (or other related data).

File: src/main/resources/migrations/common/V2024.08.19.1__reset_sequences.sql [hunk 6]
```diff
@@ -2,12 +2,12 @@
...
-alter sequence if exists contact_identity_contact_identity_id_seq restart with 4;
-alter sequence if exists contact_phone_contact_phone_id_seq restart with 12;
-alter sequence if exists reference_codes_reference_code_id_seq restart with 82;
-alter sequence if exists prisoner_contact_prisoner_contact_id_seq restart with 29;
+alter sequence if exists contact_contact_id_seq restart with 18;
+alter sequence if exists contact_address_contact_address_id_seq restart with 17;
+alter sequence if exists contact_email_contact_email_id_seq restart with 5;
+alter sequence if exists contact_identity_contact_identity_id_seq restart with 5;
+alter sequence if exists contact_phone_contact_phone_id_seq restart with 13;
+alter sequence if exists reference_codes_reference_code_id_seq restart with 83;
```
Comment (timharrison-moj) [inline src/main/resources/migrations/common/V2024.08.19.1__reset_sequences.sql:10]: Didn't need to update all the sequence numbers - just the one for contacts, where you added 3 more rows. The others are still the same.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppscontactsapi/model/request/ContactSearchResultItemRequestTest.kt [hunk 7]
```diff
@@ -6,7 +6,7 @@ import org.assertj.core.api.Assertions.assertThat
 import org.junit.jupiter.api.Test
 import java.time.LocalDate
 
-class ContactSearchRequestTest {
+class ContactSearchResultItemRequestTest {
```
Comment (timharrison-moj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppscontactsapi/model/request/ContactSearchResultItemRequestTest.kt:10]: Think this has been caught up in the name change, and probably shouldn't have? Its testing the validation of requests - not the result items. `ContactSearchRequestValidationTest` or something like that would be better.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppscontactsapi/service/ContactServiceTest.kt [hunk 8]
```diff
@@ -360,10 +360,9 @@ class ContactServiceTest {
     fun `test searchContacts with lastName , firstName , middleName and date of birth`() {
       // Given
       val pageable = PageRequest.of(0, 10)
-      val contact = getContactEntity(1L)
-      val contactAddress = getContactAddressEntity(1L, 1L)
+      val contactAddressSummaryEntity = getContactAddressSummaryEntity()
 
-      val results = listOf(arrayOf(contact, contactAddress))
+      val results = listOf(contactAddressSummaryEntity)
```
Comment (timharrison-moj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppscontactsapi/service/ContactServiceTest.kt:366]: Missed a rename? contactWithAddressEntity?

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppscontactsapi/service/ContactServiceTest.kt [hunk 9]
```diff
@@ -388,21 +387,15 @@ class ContactServiceTest {
       assertThat(result.content[0].firstName).isEqualTo("first")
     }
 
-    private fun getContactEntity(contactId: Long) = ContactEntity(
-      contactId = contactId,
+    private fun getContactAddressSummaryEntity() = ContactWithAddressEntity(
```
Comment (timharrison-moj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppscontactsapi/service/ContactServiceTest.kt:390]: Missed a rename?

**Review by timharrison-moj — COMMENTED** [review]
Nearly there .. one or two tidying bits to go.

**Review by timharrison-moj — APPROVED** [review]
LGTM.

**Pull request merged**